Test suite failure on non-AVX512 due to glibc (version-dependent) bug on f64 underflow

While [trying](https://github.com/conda-forge/numpy-feedstock/pull/235) to build 1.21.0rc1 for conda-forge, a new bug appeared, for agents without AVX512 support.

<details>
<summary>Test suite failure for <code>TestSpecialFloats.test_exp_values</code></summary>

```
=================================== FAILURES ===================================
______________________ TestSpecialFloats.test_exp_values _______________________

self = <numpy.core.tests.test_umath.TestSpecialFloats object at 0x7f1a528bb820>

    def test_exp_values(self):
        x = [np.nan,  np.nan, np.inf, 0.]
        y = [np.nan, -np.nan, np.inf, -np.inf]
        for dt in ['f', 'd', 'g']:
            xf = np.array(x, dtype=dt)
            yf = np.array(y, dtype=dt)
            assert_equal(np.exp(yf), xf)
    
        with np.errstate(over='raise'):
            assert_raises(FloatingPointError, np.exp, np.float32(100.))
            assert_raises(FloatingPointError, np.exp, np.float32(1E19))
            assert_raises(FloatingPointError, np.exp, np.float64(800.))
            assert_raises(FloatingPointError, np.exp, np.float64(1E19))
    
        with np.errstate(under='raise'):
            assert_raises(FloatingPointError, np.exp, np.float32(-1000.))
            assert_raises(FloatingPointError, np.exp, np.float32(-1E19))
>           assert_raises(FloatingPointError, np.exp, np.float64(-1000.))

../[...]/lib/python3.9/site-packages/numpy/core/tests/test_umath.py:998: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../[...]/lib/python3.9/unittest/case.py:733: in assertRaises
    return context.handle('assertRaises', args, kwargs)
../[...]/lib/python3.9/unittest/case.py:201: in handle
    callable_obj(*args, **kwargs)
../[...]/lib/python3.9/unittest/case.py:223: in __exit__
    self._raiseFailure("{} not raised by {}".format(exc_name,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <unittest.case._AssertRaisesContext object at 0x7f1a528bba90>
standardMsg = 'FloatingPointError not raised by exp'

    def _raiseFailure(self, standardMsg):
        msg = self.test_case._formatMessage(self.msg, standardMsg)
E       AssertionError: FloatingPointError not raised by exp

../[...]/lib/python3.9/unittest/case.py:163: AssertionError
=============================== warnings summary ===============================
```

</details>

@seiko2plus [commented](https://github.com/conda-forge/numpy-feedstock/pull/235#issuecomment-849238234):

> 
> It makes sense now the error disappears when the machine has AVX512 support, it seems we are dealing with compiler/libc that doesn't respect underflow error since the current exp/f64 SIMD kernel only optimizes CPUs with AVX512 Skylake features.
> 
> [...]
>
> However, Could you try the following patch?
>
> <details>
> 
> ```diff
> index 41e0bf37b..89798f74a 100644
> --- a/numpy/core/src/umath/loops_exponent_log.dispatch.c.src
> +++ b/numpy/core/src/umath/loops_exponent_log.dispatch.c.src
> @@ -1248,6 +1248,7 @@ NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(FLOAT_@func@)
>  /**begin repeat
>   * #func = exp, log#
>   * #scalar = npy_exp, npy_log#
> + * #underflow = 1, 0#
>   */
>  NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(DOUBLE_@func@)
>  (char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(data))
> @@ -1260,6 +1261,13 @@ NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(DOUBLE_@func@)
>  #endif
>      UNARY_LOOP {
>          const npy_double in1 = *(npy_double *)ip1;
> +    #if @underflow@
> +        if (NPY_UNLIKELY(!npy_isnan(in1) && in1 < -0x1.74910d52d3053p+9)) {
> +            *(npy_double *)op1 = 0;
> +            npy_set_floatstatus_underflow();
> +            continue;
> +        }
> +    #endif
>          *(npy_double *)op1 = @scalar@(in1);
>      }
>  }
> ```
>
> </details>
> 
> if the patch passes the test, then we will need to determine which versions of glibc are affected by this bug.

And indeed, the test passes with that patch, which is why I'm opening this issue.

Sidenote: based on the mention of (g)libc, this might be somewhat similar to https://github.com/numpy/numpy/issues/15179? The corresponding test is currently being skipped in conda-forge, but _does_ pass with the following patch (suggested by @r-devulap):

<details>

```diff
diff --git a/numpy/core/tests/test_umath.py b/numpy/core/tests/test_umath.py
index 9d1b13b53..faa2ca8f0 100644
--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -1188,8 +1188,6 @@ def test_sincos_float32(self):
         M = np.int_(N/20)
         index = np.random.randint(low=0, high=N, size=M)
         x_f32 = np.float32(np.random.uniform(low=-100.,high=100.,size=N))
-        # test coverage for elements > 117435.992f for which glibc is used
-        x_f32[index] = np.float32(10E+10*np.random.rand(M))
         x_f64 = np.float64(x_f32)
         assert_array_max_ulp(np.sin(x_f32), np.float32(np.sin(x_f64)), maxulp=2)
         assert_array_max_ulp(np.cos(x_f32), np.float32(np.cos(x_f64)), maxulp=2
```

</details>

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -19,11 +19,10 @@ from numpy.testing import (
     )
 
 def get_glibc_version():
-    ver = '0.0'
     try:
         ver = os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1]
     except Exception as inst:
-        print(inst)
+        ver = '0.0'
 
     return ver
 

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -19,13 +19,18 @@ from numpy.testing import (
     )
 
 def get_glibc_version():
-    ver = 0.0
+    ver = '0.0'
     try:
-        ver = float(os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1])
+        ver = os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1]
     except Exception as inst:
         print(inst)
 
     return ver
+
+glibcver = get_glibc_version()
+glibc_newerthan_2_17 = pytest.mark.xfail(
+        glibcver != '0.0' and glibcver < '2.17',
+        reason="Older glibc versions may not raise appropriate FP exceptions")
 
 def on_powerpc():
     """ True if we are running on a Power PC platform."""
@@ -996,20 +1001,21 @@ class TestSpecialFloats:
             yf = np.array(y, dtype=dt)
             assert_equal(np.exp(yf), xf)
 
-        # Older version of glibc may not raise the correct FP exceptions
-        # See: https://github.com/numpy/numpy/issues/19192
-        if (get_glibc_version() >= 2.17):
-            with np.errstate(over='raise'):
-                assert_raises(FloatingPointError, np.exp, np.float32(100.))
-                assert_raises(FloatingPointError, np.exp, np.float32(1E19))
-                assert_raises(FloatingPointError, np.exp, np.float64(800.))
-                assert_raises(FloatingPointError, np.exp, np.float64(1E19))
+    # Older version of glibc may not raise the correct FP exceptions
+    # See: https://github.com/numpy/numpy/issues/19192
+    @glibc_newerthan_2_17
+    def test_exp_exceptions(self):
+        with np.errstate(over='raise'):
+            assert_raises(FloatingPointError, np.exp, np.float32(100.))
+            assert_raises(FloatingPointError, np.exp, np.float32(1E19))
+            assert_raises(FloatingPointError, np.exp, np.float64(800.))
+            assert_raises(FloatingPointError, np.exp, np.float64(1E19))
 
-            with np.errstate(under='raise'):
-                assert_raises(FloatingPointError, np.exp, np.float32(-1000.))
-                assert_raises(FloatingPointError, np.exp, np.float32(-1E19))
-                assert_raises(FloatingPointError, np.exp, np.float64(-1000.))
-                assert_raises(FloatingPointError, np.exp, np.float64(-1E19))
+        with np.errstate(under='raise'):
+            assert_raises(FloatingPointError, np.exp, np.float32(-1000.))
+            assert_raises(FloatingPointError, np.exp, np.float32(-1E19))
+            assert_raises(FloatingPointError, np.exp, np.float64(-1000.))
+            assert_raises(FloatingPointError, np.exp, np.float64(-1E19))
 
     def test_log_values(self):
         with np.errstate(all='ignore'):

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -22,8 +22,8 @@ def get_glibc_version():
     ver = 0.0
     try:
         ver = float(os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1])
-    except:
-        print("Couldn't capture GLIBC version")
+    except Exception as inst:
+        print(inst)
 
     return ver
 

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -27,6 +27,7 @@ def get_glibc_version():
 
     return ver
 
+
 glibcver = get_glibc_version()
 glibc_newerthan_2_17 = pytest.mark.xfail(
         glibcver != '0.0' and glibcver < '2.17',

--- a/numpy/core/tests/test_umath.py
+++ b/numpy/core/tests/test_umath.py
@@ -4,6 +4,7 @@ import fnmatch
 import itertools
 import pytest
 import sys
+import os
 from fractions import Fraction
 from functools import reduce
 
@@ -16,6 +17,15 @@ from numpy.testing import (
     assert_array_max_ulp, assert_allclose, assert_no_warnings, suppress_warnings,
     _gen_alignment_data, assert_array_almost_equal_nulp, assert_warns
     )
+
+def get_glibc_version():
+    ver = 0.0
+    try:
+        ver = float(os.confstr('CS_GNU_LIBC_VERSION').rsplit(' ')[1])
+    except:
+        print("Couldn't capture GLIBC version")
+
+    return ver
 
 def on_powerpc():
     """ True if we are running on a Power PC platform."""
@@ -986,17 +996,20 @@ class TestSpecialFloats:
             yf = np.array(y, dtype=dt)
             assert_equal(np.exp(yf), xf)
 
-        with np.errstate(over='raise'):
-            assert_raises(FloatingPointError, np.exp, np.float32(100.))
-            assert_raises(FloatingPointError, np.exp, np.float32(1E19))
-            assert_raises(FloatingPointError, np.exp, np.float64(800.))
-            assert_raises(FloatingPointError, np.exp, np.float64(1E19))
+        # Older version of glibc may not raise the correct FP exceptions
+        # See: https://github.com/numpy/numpy/issues/19192
+        if (get_glibc_version() >= 2.17):
+            with np.errstate(over='raise'):
+                assert_raises(FloatingPointError, np.exp, np.float32(100.))
+                assert_raises(FloatingPointError, np.exp, np.float32(1E19))
+                assert_raises(FloatingPointError, np.exp, np.float64(800.))
+                assert_raises(FloatingPointError, np.exp, np.float64(1E19))
 
-        with np.errstate(under='raise'):
-            assert_raises(FloatingPointError, np.exp, np.float32(-1000.))
-            assert_raises(FloatingPointError, np.exp, np.float32(-1E19))
-            assert_raises(FloatingPointError, np.exp, np.float64(-1000.))
-            assert_raises(FloatingPointError, np.exp, np.float64(-1E19))
+            with np.errstate(under='raise'):
+                assert_raises(FloatingPointError, np.exp, np.float32(-1000.))
+                assert_raises(FloatingPointError, np.exp, np.float32(-1E19))
+                assert_raises(FloatingPointError, np.exp, np.float64(-1000.))
+                assert_raises(FloatingPointError, np.exp, np.float64(-1E19))
 
     def test_log_values(self):
         with np.errstate(all='ignore'):

